pkg/agent/agent.go

```diff
@@ -663,6 +674,62 @@ func (c *agentController) SyncAddOnPlacementScore(ctx context.Context) error {
 	return nil
 }
 
+func (c *agentController) deleteManagedCluster(ctx context.Context, hc *hyperv1beta1.HostedCluster) error {
+	if hc == nil {
+		return fmt.Errorf("failed to delete nil hostedCluster")
+	}
+
+	managedClusterName, ok := hc.GetAnnotations()[util.ManagedClusterAnnoKey]
+	if !ok || len(managedClusterName) == 0 {
+		managedClusterName = hc.Name
+	}
+
+	klusterletName := "klusterlet-" + managedClusterName
+
+	// Remove the operator.open-cluster-management.io/klusterlet-hosted-cleanup finalizer in klusterlet
+	klusterlet := &operatorapiv1.Klusterlet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: klusterletName,
+		},
+	}
+
+	if err := c.spokeClient.Get(ctx, client.ObjectKeyFromObject(klusterlet), klusterlet); err != nil {
+		if apierrors.IsNotFound(err) {
+			c.log.Info(fmt.Sprintf("klusterlet %v is already deleted", klusterletName))
+		} else {
+			c.log.Error(err, fmt.Sprintf("failed to get the klusterlet %v", klusterletName))
+			return err
+		}
+	}
+
+	updated := controllerutil.RemoveFinalizer(klusterlet, "operator.open-cluster-management.io/klusterlet-hosted-cleanup")
+	c.log.Info(fmt.Sprintf("klusterlet %v finalizer removed:%v", klusterletName, updated))
+
+	// Delete the managed cluster
+	mc := &clusterv1.ManagedCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: managedClusterName,
+		},
+	}
+
+	if err := c.spokeClient.Get(ctx, client.ObjectKeyFromObject(mc), mc); err != nil {
+		if apierrors.IsNotFound(err) {
+			c.log.Info(fmt.Sprintf("managedCluster %v is already deleted", managedClusterName))
+		} else {
+			c.log.Error(err, fmt.Sprintf("failed to get the managedCluster %v", managedClusterName))
+			return err
+		}
+	}
+
+	if err := c.spokeClient.Delete(ctx, mc); err != nil {
```

Comment:
I guess we should not use the spoke client to delete the managedCluster CR, since the managedCluster CR is on the hub cluster.